dfir_rs/src/scheduled/metrics.rs

```diff
@@ -0,0 +1,192 @@
+//! Runtime metrics for DFIR.
+
+use std::cell::Cell;
+use std::iter::FusedIterator;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::task::{Context, Poll};
+
+use pin_project_lite::pin_project;
+use web_time::{Duration, Instant};
+
+use super::{HandoffId, HandoffTag, SubgraphId, SubgraphTag};
+use crate::util::slot_vec::SecondarySlotVec;
+
+#[derive(Default, Clone)]
+pub(super) struct DfirMetricsState {
+    pub(super) subgraph_metrics: SecondarySlotVec<SubgraphTag, SubgraphMetrics>,
+    pub(super) handoff_metrics: SecondarySlotVec<HandoffTag, HandoffMetrics>,
+}
+
+/// DFIR runtime metrics accumulated across a time, possibly since runtime creation.
+#[derive(Clone)]
+pub struct DfirMetrics {
+    pub(super) curr: Rc<DfirMetricsState>,
+    /// `None` for "since creation".
+    pub(super) prev: Option<DfirMetricsState>,
+}
+impl DfirMetrics {
+    /// Begins a new metrics collection period, effectively resetting all metrics to zero.
+    pub fn reset(&mut self) {
+        self.prev = Some(self.curr.as_ref().clone());
+    }
+
+    /// Returns an iterator over all subgraph IDs.
+    pub fn subgraph_ids(
+        &self,
+    ) -> impl '_ + DoubleEndedIterator<Item = SubgraphId> + FusedIterator + Clone {
+        self.curr.subgraph_metrics.keys()
+    }
+
+    /// Gets the metrics for a particular subgraph.
```

Comment:
Is it standard with the metrics packages you're dumping to to report delta numbers like this as opposed to cumulative numbers?